php/libraries/File_Upload_Validator.class.inc

```diff
@@ -0,0 +1,177 @@
+            // catch this exception.
+            $this->_responseCode = 403;
+            throw new \LorisException(
+                'Upload directory is not writable!'
+            );
+        }
+        return;
+    }
+
+    /**
+     * Given an array of files, validate them and upload them to the server.
+     *
+     * @param array $uploadedFiles An array of files uploaded to the server.
+     *                              These files must implement the PSR-7
+     *                              UploadedFile interface.
+     * @param int   $permissions   A UNIX permissions string (e.g. 644). All files
+     *                              uploaded will have this permission string.
+     *                              If uploaded files are required to have
+     *                              differing permissions, this must be done
+     *                              manually in the calling code after this
+     *                              functions has executed.
+     *
+     * @return bool Whether the file upload was successful. Modifies
+     *                      $this->__responseCode as a side-effect.
+     */
+    public function upload(
+        array $uploadedFiles,
+        int $permissions
+    ) : bool {
+        if (empty($uploadedFiles)) {
+            return false;
+        }
+
+        foreach ($uploadedFiles as $uploadedFile) {
+            //Ensure that the $uploadedFile uses the PSR-7 UploadedFile
+            //interface
+            if (! $uploadedFile instanceof \Zend\Diactoros\UploadedFile) {
+
+                error_log(
+                    'File uploaded to File_Upload_Validator is not an ' .
+                    'instance of PSR-7 UploadedFIle. Aborting upload.'
+                );
+                $this->_responseCode = 500;
+                return false;
+            }
+            /* basename() is used here to prevent path traversal characters
+             * from being used.
+             */
+            $targetPath = $this->uploadDirectory . basename(
+                $uploadedFile->getClientFilename()
+            );
+            // If file exists, set response code to 'Conflict'
+            if (file_exists($targetPath)) {
+                $this->_responseCode = 409;
+                return false;
+            }
+            // Validate theat the MIME type of the file is allowed
+            if (!empty($this->permittedMIMETypes)) {
+                if (!in_array(
+                    $uploadedFile->getClientMediaType(),
+                    $this->permittedMIMETypes,
+                    true
+                )
+                ) {
+                    $this->_responseCode = 400;
+                    return false;
+                }
+            }
+
+            // Try to move the file to $targetPath
+            try {
+                $uploadedFile->moveTo($targetPath);
+            } catch (\Exception $e) {
+                /* If an error occurs during moving, set response code to
```

Comment:
An exception is not an error. An exception is an exception. I don't think this should be caught.